File: magpie/src/pixel_utils.f90

subroutine which_pix_id_scalar(x, xmin, dx, pix_id)

  ! Find pixel along a defined grid.
  !
  ! Parameters
  ! ----------
  ! x : float
  !   X-coordinates.
  ! xmin : float
  !   Minimum along the grid.
  ! dx : float
  !   pixel width.
  !
  ! Returns
  ! -------
  ! pix_id : int
  !   The pixel the points are located in.

  implicit none

  ! Parameter declarations

  integer, parameter :: dp = kind(1.d0)

  real(kind=dp), intent(in) :: x, xmin, dx
  integer, intent(out) :: pix_id

  ! Main

  pix_id = int(floor((x-xmin)/dx))

end subroutine which_pix_id_scalar


subroutine which_pix_id_array(x, xmin, dx, xlen, pix_id)

  ! Find pixel along a defined grid.
  !
  ! Parameters
  ! ----------
  ! x : array
  !   X-coordinates.
  ! xmin : float
  !   Minimum along the grid.
  ! dx : float
  !   pixel width.
  !
  ! Returns
  ! -------
  ! pix_id : int
  !   The pixel the points are located in.

  implicit none

  ! Parameter declarations

  integer, parameter :: dp = kind(1.d0)
  integer, intent(in) :: xlen
  real(kind=dp), intent(in) :: x(xlen), xmin, dx
  integer, intent(out) :: pix_id(xlen)
  integer :: i

  ! Main

  do i = 1, xlen
    call which_pix_id_scalar(x(i), xmin, dx, pix_id(i))
  end do

end subroutine which_pix_id_array

File: magpie/src/pixel_utils.py
import numpy as np
from numba import njit


@njit
def which_pix_id_scalar(x, xmin, dx):
    """
    Finds the pixel index along a defined grid.

    Parameters
    ----------
    x : float
        X-coordinate.
    xmin : float
        Minimum along the grid.
    dx : float
        Pixel width.

    Returns
    -------
    pix_id : int
        The pixel index in which the point is located.
    """
    return int(np.floor((x - xmin) / dx))


@njit
def which_pix_id_array(x, xmin, dx):
    """
    Finds the pixel indices for an array of x-coordinates along a defined grid.

    Parameters
    ----------
    x : array-like (float)
        X-coordinates.
    xmin : float
        Minimum along the grid.
    dx : float
        Pixel width.

    Returns
    -------
    pix_id : array-like (int)
        The pixel indices corresponding to each x value.
    """
    pix_id = np.empty(len(x), dtype=np.int32)
    for i in range(len(x)):
        pix_id[i] = which_pix_id_scalar(x[i], xmin, dx)
    return pix_id
